### PR TITLE
Fix #934 - Modify TripService to start service in foreground

### LINKED ISSUE #934: Reminders don't reschedule after reboot when targeting Android 8.0 (API level 26)
**Summary:** 

This issue has the same underlying cause as https://github.com/OneBusAway/onebusaway-android/issues/933 - this crash is related to background executions limits introduced on Oreo:
https://developer.android.com/about/versions/oreo/background

If you schedule an arrival reminder and then restart your device, you'll get this stacktrace:

~~~
    Process: com.joulespersecond.seattlebusbot, PID: 4718
    java.lang.RuntimeException: Unable to start receiver org.onebusaway.android.tripservice.BootstrapService: java.lang.IllegalStateException: 
	Not allowed to start service Intent { act=com.joulespersecond.seattlebusbot.action.SCHEDULE dat=content://com.joulespersecond.oba/trips cmp=com
	.joulespersecond.seattlebusbot/org.onebusaway.android.tripservice.TripService }: 
	app is in background uid UidRecord{b5c9138 u0a176 RCVR idle change:idle|uncached procs:1 seq(0,0,0)}
        at android.app.ActivityThread.handleReceiver(ActivityThread.java:3523)
        at android.app.ActivityThread.access$1400(ActivityThread.java:207)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1759)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6863)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.lang.IllegalStateException: Not allowed to start service Intent { act=com.joulespersecond.seattlebusbot.action.SCHEDULE 
	 dat=content://com.joulespersecond.oba/trips cmp=com.joulespersecond.seattlebusbot/org.onebusaway.android.tripservice.TripService }: 
	 app is in background uid UidRecord{b5c9138 u0a176 RCVR idle change:idle|uncached procs:1 seq(0,0,0)}
        at android.app.ContextImpl.startServiceCommon(ContextImpl.java:1595)
        at android.app.ContextImpl.startService(ContextImpl.java:1550)
        at android.content.ContextWrapper.startService(ContextWrapper.java:664)
        at android.content.ContextWrapper.startService(ContextWrapper.java:664)
        at org.onebusaway.android.tripservice.TripService.scheduleAll(TripService.java:231)
        at org.onebusaway.android.tripservice.BootstrapService.onReceive(BootstrapService.java:33)
        at android.app.ActivityThread.handleReceiver(ActivityThread.java:3507)
        at android.app.ActivityThread.access$1400(ActivityThread.java:207) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1759) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:193) 
        at android.app.ActivityThread.main(ActivityThread.java:6863) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:537) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858) 
~~~

Our `BootstrapService` (which is actually a `BroadcastReceiver`) is triggered by a `android.intent.action.BOOT_COMPLETED` Intent on reboot, which then tries to reschedule the reminders by running the TripService with the `com.joulespersecond.seattlebusbot.action.SCHEDULE` Intent.

However, even though [the docs](https://developer.android.com/about/versions/oreo/background) say:

>Under certain circumstances, a background app is placed on a temporary whitelist for several minutes. While an app is on the whitelist, it can launch services without limitation, and its background services are permitted to run. An app is placed on the whitelist when it handles a task that's visible to the user, such as:
>
> * Handling a high-priority Firebase Cloud Messaging (FCM) message.
> * Receiving a broadcast, such as an SMS/MMS message.
> * Executing a PendingIntent from a notification.
> * Starting a VpnService before the VPN app promotes itself to the foreground.

...apparently the "Receiving a broadcast" exception above doesn't include the `android.intent.action.BOOT_COMPLETED` broadcast, because we still get the IllegalStateException.

This StackOverflow post summarizes the issue, and the accepted answer outlines a few options:
https://stackoverflow.com/questions/44502229/runtime-exception-android-o-with-boot-completed

@paulnabanita1 is going to test starting the TripService in the foreground (Option 1) to see if this avoids the `IllegalStateException`.  I've heard from other developers that switching to foreground services avoids a lot of headaches with unexpected background behavior.

@paulnabanita1 Note that if we implement a full solution based on the foreground execution, we'll need to customize the notification message based on what the TripService is doing (e.g., Scheduling reminders, checking for new real-time info from the server, etc.), which would be in `TripService.handleCommand()` based on the value of `intent.getAction()` (i.e., the IF statements starting on line 181).

To ensure that the reboot broadcast is received on all devices, as mentioned [here](https://github.com/AdguardTeam/AdguardForAndroid/issues/183) we should also expand the `intent-filter` in the `AndroidManifest.xml` to include the following `QUICKBOOT_POWERON` intents:

~~~
            <intent-filter>
                <action android:name="android.intent.action.BOOT_COMPLETED"/>
                <action android:name="android.intent.action.QUICKBOOT_POWERON"/>
                <action android:name="com.htc.intent.action.QUICKBOOT_POWERON"/>
            </intent-filter>
~~~

Apparently these are triggered when choosing a device restart, vs. powering down and powering back on (we could also consider including `ACTION_LOCKED_BOOT_COMPLETED` - https://developer.android.com/reference/android/content/Intent.html#ACTION_LOCKED_BOOT_COMPLETED - but for now let's just include the ones above).

Based on these references:
* https://github.com/OpenLocate/openlocate-android/issues/70
* https://issuetracker.google.com/issues/76112072
* https://github.com/AdguardTeam/AdguardForAndroid/issues/183

Android docs for starting the service in the foreground:
*https://developer.android.com/reference/android/app/Service.html#startForeground(int,%20android.app.Notification)

**Steps to reproduce:** 
1. Run app using code on the master branch (target SDK 26)
1. Schedule a reminder
1. Restart your device

**Expected behavior:** 

TripService should be started and reschedule any reminders

**Observed behavior:** 

When trying to start TripService in the background it crashes with an IllegalStateException.

**Device and Android version:** 

OnePlus 5T with Android 9.0 Pie

(@paulnabanita1 please let me know if I got the above make/model of your device right)

### PR DESCRIPTION
**Pull request by paulnabanita1**
Modified TripService.java to mitigate the issue with nothing getting bus arrival notifications after device restart.

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [ ] Run the unit tests with `gradlew connectedObaGoogleDebugAndroidTest connectedObaAmazonDebugAndroidTest` to make sure you didn't break anything

- [ ] If you have multiple commits please combine them into one commit by squashing them for the initial submission of the pull request.  When addressing comments on a pull request, please push a new commit per comment when possible (reviewers will squash and merge using GitHub merge tool)

### CODE REVIEW

File: onebusaway-android/src/main/java/org/onebusaway/android/tripservice/TripService.java [hunk 1]
```diff
@@ -125,6 +129,20 @@ public void onStart(Intent intent, int startId) {
...
+        PendingIntent pendingIntent = PendingIntent.getActivity(this, 0,
+                notificationIntent, 0);
+
+        Notification notification = new NotificationCompat.Builder(getApplicationContext(), Application.CHANNEL_TRIP_PLAN_UPDATES_ID)
+                .setSmallIcon(R.drawable.ic_stat_notification)
+                .setContentTitle("Service running in foreground")
+                .setContentText("OneBusAway will be running a service in foreground")
+                .setContentIntent(pendingIntent).build();
+
+        startForeground(1001,notification);
```
Comment (barbeau) [inline onebusaway-android/src/main/java/org/onebusaway/android/tripservice/TripService.java]: Could you change the ID here to a larger number like `1000000`, and refactor it to a `public static final` member constant in the class?  We don't want this to collide with other notification IDs we use in OBA.  A large number seems the best shot for this to me, but let me know if you have any other ideas.

Comment (paulnabanita1) [reply onebusaway-android/src/main/java/org/onebusaway/android/tripservice/TripService.java]: Thank you for reviewing the PR. I will add the changes and test if it works as expected.

Comment (paulnabanita1) [reply onebusaway-android/src/main/java/org/onebusaway/android/tripservice/TripService.java]: @barbeau I have updated the PR. Can you please review it and let me know if any further changes are required to be made?

Comment (paulnabanita1) [reply onebusaway-android/src/main/java/org/onebusaway/android/tripservice/TripService.java]: @barbeau Thank you for updating the PR. I have tested the changes and everything is working as expected.

**Pull request merged**